### ISSUE TITLE
tf.contrib.graph_editor Shape error

### ISSUE THREAD

**Issue by jrbtaylor** [tensorflow/tensorflow#20026]
Using tf.contrib.graph_editor for pruning, specifically the graph_replace function, raises ValueErrors.

As a trivial example, consider a 2-layer network with weight matrices W0 [3,4] and W1 [4,1] that we want to prune one node from and replace with new weight matrices W0' [3,3] and W1' [3,1]. From the documentation, this appears be straight-forward to do using graph_replace and the resulting graph in theory shouldn't have any shape incompatibilities.

The documentation should be updated to clarify this limitation and a proper function for use with pruning should be identified if it exists.

**Comment by jrbtaylor** [comment]
@purpledog can you please clarify if there's another function in graph_editor that could be used for network pruning?

**Comment by purpledog** [comment]
From the description you gave it should be fine. would you mind sending some code?

**Comment by jrbtaylor** [comment]
It appears the problem is in the copy_op_handler function:
`def copy_op_handler(info, op, new_inputs, copy_shape=True):`
I changed the default copy_shape to False and now it works. None of the calls to copy_op_handler seem to pass this argument though - is there a reason why it defaults to True?

**Comment by jrbtaylor** [comment]
Here's a self-contained example:

```
import numpy as np
import tensorflow as tf
from tensorflow.contrib import graph_editor as ge

def prune_test():
    # make a simple graph
    x = tf.placeholder(tf.float32, shape=[1, 2], name='x')
    w0 = tf.Variable(np.array([[1, 2, 3, 4], [6, 7, 8, 9]], dtype=np.float32), name='w0')
    x1 = tf.matmul(x, w0, name='x1')
    w1 = tf.Variable(np.array([[-1], [2], [-3], [4]], dtype=np.float32), name='w1')
    y = tf.matmul(x1, w1, name='y')

    def vars_to_tensors(v_list):
        def rsub(string, sub):
            return string[:len(string)-len(sub)] if string.endswith(sub) else string
        graph = tf.get_default_graph()
        nodes = [n for n in graph.as_graph_def().node if n.op == 'VariableV2']
        nodes = [next(n for n in nodes if rsub(v.name, ':0') == rsub(n.name, ':0')) for v in v_list]
        return [graph.get_tensor_by_name(node.name + ':0') for node in nodes]

    w0p = tf.Variable(np.array([[1, 2, 3], [6, 7, 8]], dtype=np.float32), name='w0_pruned')
    w1p = tf.Variable(np.array([[-1], [2], [-3]], dtype=np.float32), name='w1_pruned')

    # get tensors since graph_replace doesn't work with variables directly
    [w0, w1, w0p, w1p] = vars_to_tensors([w0, w1, w0p, w1p])
    replacements = {w0: w0p, w1: w1p}
    yp = ge.graph_replace(y, replacements)

    with tf.Session() as sess:
        sess.run(tf.global_variables_initializer())
        output, pruned_output = sess.run([y, yp], feed_dict={x: [[4., 7.]]})
    print(output)
    print(pruned_output)
```

With the default value for copy_shape in copy_op_handler set to True, this fails with a shape error. With it set to False, this runs as expected.

**Comment by jrbtaylor** [comment]
I added a [pull request](https://github.com/tensorflow/tensorflow/pull/20039) for this if you think the default value should be changed @purpledog 
I may be missing some situation where copying the shape is preferable though. Perhaps you could weigh in on that use case.

**Comment by purpledog** [comment]
Thanks Jason. This flags was useful at some point two years ago but I can't remember why and it might not be useful anymore. Your change is very much welcome. If it passes the test and make your life easier, then so be it!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @purpledog: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @purpledog: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @purpledog: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @purpledog: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jrbtaylor** [comment]
resolved by #20039
